pkg/endpoint/bpf.go

```diff
@@ -604,21 +612,20 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
 		// GH-3897 would fix this by creating a new map to do an atomic swap
 		// with the old one.
 		stats.mapSync.Start()
-		err := e.syncPolicyMap()
-		stats.mapSync.End(err == nil)
+		err = e.syncPolicyMap()
 		if err != nil {
-			e.Unlock()
-			return 0, compilationExecuted, fmt.Errorf("unable to regenerate policy because PolicyMap synchronization failed: %s", err)
+			return proxyCleanupFunc, fmt.Errorf("unable to regenerate policy because PolicyMap synchronization failed: %s", err)
 		}
+		stats.mapSync.End(err == nil)
 
 		// Configure the new network policy with the proxies.
 		stats.proxyPolicyCalculation.Start()
 		var networkPolicyRevertFunc revert.RevertFunc
-		err, networkPolicyRevertFunc = e.updateNetworkPolicy(owner, proxyWaitGroup)
+		err, networkPolicyRevertFunc = e.updateNetworkPolicy(owner, datapathRegenCtx.proxyWaitGroup)
 		stats.proxyPolicyCalculation.End(err == nil)
 		if err != nil {
 			e.Unlock()
```

Comment:
This should be removed, right? the deferred one will do the right thing. There are a few more below as well.

Reply:
The deferred function performs any cleanup or finalization for the revert / finalize stacks, not actual plumbing of the network policy for the proxy.